DSESwift/DataStorageEngine.swift

```diff
@@ -70,4 +165,121 @@ public class DataStorageEngine{
     public func getSources() -> [String]{
         return DatabaseHandler.getSources()
     }
+    
+    /**
+     * Returns enum DSEStatus indicating status of DSE.
+     * @return The DSEStatus, this could be either AWAITING_CREDENTIALS, AWAITING_SENSOR_PROFILES, READY.
+     **/
+    public func getStatus() -> DSEStatus{
+        if(self.config.sessionId == nil) {
+            return DSEStatus.AWAITING_CREDENTIALS;
+        } else if(self.dataSyncer.initialized) {
+            return DSEStatus.READY;
+        } else {
+            return DSEStatus.AWAITING_SENSOR_PROFILES;
+        }
+    }
+
+    /**
+    * Synchronizes the local data with Common Sense asynchronously
+    * The results will be returned via AsyncCallback
+    **/
+    func syncData(completionHandler : DSEAsyncCallback?) throws {
+        if (completionHandler != nil){
+            syncCompleteHandlers.append(completionHandler!)
+        }
+        self.dataSyncer.sync()
+    }
+    
+    /**
+    * Notifies when the sensors are downloaded asynchronously
+    * The sensors are automatically downloaded when the DataStorageEngine is ready
+    * @param callback The AsyncCallback method to call the success function on when the sensors are downloaded
+    **/
+    func onSensorsDownloaded(callback: DSEAsyncCallback){
+        if self.isSensorDownloadCompleted{
+            callback.onSuccess()
+        } else {
+            self.sensorsDownloadedCallbacks.append(callback)
+        }
+    }
+    
+    /**
+    * Notifies when the sensors are downloaded asynchronously
+    * The sensor data is automatically downloaded when the DataStorageEngine is ready
+    * @param callback The AsyncCallback method to call the success function on when the sensor data is downloaded
+    **/
+    func onSensorDataDownloaded(callback: DSEAsyncCallback){
+        if self.isSensorDataDownloadCompleted{
+            callback.onSuccess()
+        } else {
+            self.sensorDataDownloadedCallbacks.append(callback)
+        }
+    }
+    
+    /**
+    * Notifies when the initialization is done asynchronously
+    * The initialization is done when the credentials have been set and sensor profiles downloaded.
+    * @param callback The AsyncCallback method to call the success function on when ready
+    **/
+    func onReady(callback : DSEAsyncCallback){
+        if (getStatus() == DSEStatus.READY) {
+            callback.onSuccess()
+        } else {
+            // status not ready yet. keep the callback in the array in DataSyncer
+            self.readyCallbacks.append(callback)
```

Comment:
ready implies success whereas it could fail? initializationCallbacks maybe more fitting?